code/client/src/pages/Show/Swap.jsx

```diff
@@ -411,12 +446,21 @@ const Swap = ({ address }) => {
     }
   }
 
+  const approveToken = () => {
+    // TODO: implement approve token.
+    console.log(`Approving token [${tokenFrom.symbol}]`)
+  }
+
   const swapAllowed =
     tokenFrom.value !== '' &&
     tokenTo.value !== '' &&
     fromAmountFormatted !== '' && !isNaN(fromAmountFormatted) &&
     toAmountFormatted !== '' && !isNaN(toAmountFormatted)
 
+  const tokenApproved = !tokenAllowance || (tokenAllowance && tokenAllowance.gt(fromAmount || new BN(0)))
+
+  const insufficientLiquidity = tokenReserve && (tokenReserve.reserve1 === '0' || tokenReserve.reserve2 === '0')
```

Comment:
`reserve0` and `reserve1`, also need to determine which corresponds to which (tokenFrom and tokenTo), and  making sure both amounts are no greater than the reserves